### DISCUSSION TITLE
How would a drag and drop feature work?

### DISCUSSION THREAD

**Discussion by ransome1** [ransome1/sleek#230] [Ideas]
Hi there.

A couple of times a drag and drop feature has been requested but never really described or discussed. Although I'm not going to be able to work on it asap, I would like to figure out how this feature would work in the boundaries of sleek.

I can think of two ways to approach this:

## 1.
The obvious rough idea would be to drag an item and while hovering over another item, a visual indicator would appear below that hovered item. 

And now comes the tricky part as we have an automated sorting: To make sure the item would get sorted exactly below the hovered item (which is expected by the user imo), the dragged item would need to lose specific attributes (project, context, due date, priority) and adapt those from the todo it is hovered over. This can be problematic.

## 2.
Another idea would be to only allow the drag and drop feature on the groups (priority, due date, context, project). Like this a user would drag and item and hover it over a specific group. Which would then show an indicator (maybe a frame around the whole group). After dropping the item on eg a project, the dragged todos would be extended by the groups project.

What do you think?

**Comment by Saturo78** [comment]
I think i don´t need such a feature.
I also see a high risk of changing things by mistake.

**Comment by freecicero** [comment]
I too would like this feature.  

I noticed today in using Markor on Android that although it is a pretty simple text editor, its todo.txt implementation allows the movement up and down of items in the todo.txt files by providing a simple up or down arrow in the toolbar.    Clicking that part of the toolbar moves up or down whichever item the cursor is currently inside.

Granted that's not full or even partial drag and drop, but it does provide the ability to easily move an item up and down in the todo.txt file.  That's a welcome ability, and something that I don't even think Simpletask (again Android) offers. 

I note than when you highlight an item in Sleek and left click you get a simple context menu that lets you change the priority and copy or delete.  Might be simple enough to add into that menu the ability to move the item up or down in the todo.txt file, which is what Markor does.

**Reply by aisbergde** [reply]
Instead of dragging items onto buttons, or in parallel , I could also imagine do drag and drop buttons onto items, to assign them.

And I could also imagine to drag a context or project (or whatever exists as soparate optical unit) from the item to the outside, to easy remove it from the item.

**Comment by cldjr1** [comment]
My suggestion is to allow drag and drop while maintaining automatic sorting with some simple rules. 
- To reorder the todos in the list where it wouldn't affect the todo's attributes, just use basic re-ordering: in the same way that in text editors like VSCode you can move a line up/down, we should be able to move todos in the txt file up and down. 
- Click and hold on an item to initiate drag and drop | or add a small handle to todos in the ui that can be dragged to move items.
- An up/down arrow to jog items up and down as others have mentioned would also be nice. 
- A shortuct key to move todos up and down in the list would also be nice. 

When it comes to moving between attributes and auto-sorting, some logic needs to be applied. 
- When dragging something into the area of the top-level attribute in the sorting rules, the todo can be dragged above, between, or below todos in that group. 
- When dragging something into the area of one or more lower-level attributes in the sorting rules, the todo or "sub-group" of todos with that attribute could be slightly indented or the attribute highlighted during the drag action to indicate the todo will inherit that attribute. 

The expected behavior when dragging a todo will differ based on the attribute: 
- Attributes with only one possible instance, such as priority, dates, recurrences, and pomodoros: The todo should be reassigned the attribute of the destination. 
- Attributes such as Projects and Contexts: the todo should keep any current assignment and also inherit those of the destination.

This way, auto-sorting is unaffected.

**Comment by fredofredofred** [comment]
Note - using version 1.3.1, so I just realised these comments might not be accurate for the current version!

TLDR:
- For drag and drop all of the priorities would ideally appear
- Option 2 seems like a better concept because
  - I don't like the idea of removing context to make it work, seems like making one thing less intuitive and another more.
  - Also apart from simple sequencing todo.txt is not well suited to the task, therefore I am not sure it is worth producing features tailored towards sequencing rather than general features.
___

Sleek currently only shows the priorities in use, so if I want something to be priority 'B' but I don't currently have a priority 'B' item then that would need to appear for drag and drop to work nicely. 

~~Another option might be a system where you can promote/demote items by one priority level, whcihc is often what I find myself doing. That could be a button~~ OK, I have just found this option is already there on right-click, personally that is convenient enough for how I use Sleek.

Critique of Option 1:

Option 1 is about using Sleek for sequencing, with the order within a priority able to be adjusted manually. I guess this would mean turning off sorting and changing the order of lines in the file to suit the preferred order. I am not sure how else this can be made to work with the todo.txt approach in an intuitive way - removing context seems unintuitive if I just intended to reorder things.

If someone is working out sequencing then either it's really simple and they choose to use todo.txt, or it's not and todo.txt is probably never going to be a great tool.  Sequencing is about dependencies - recording them and ideally reordering items accordingly. 

The best basic way I've found to do sequencing manually is to use lists in a text editor which includes buttons to move, promote or demote items 'with subpoints'. Often you need to be able to move an item along with all the items which depend on it to preserve the dependencies. 

I can't think of a way to do this using todo.txt without adding on additional features which wouldn't be supported by other programs.

**Reply by fredofredofred** [reply]
BTW I am using the old 1.3.1 version of Sleek so maybe  - I check the new one fairly regularly but when I checked a while back markdown and links weren't displaying as I'd like.
